hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -158,96 +169,63 @@ public Map<DatanodeDetails, SCMCommand<?>> processAndCreateCommands(
-        // We have source nodes.
-        if (sources.size() >= repConfig.getData()) {
-          final List<DatanodeDetails> selectedDatanodes = getTargetDatanodes(
-              excludedNodes, container, missingIndexes.size());
-
-          if (validatePlacement(nodes, selectedDatanodes)) {
-            excludedNodes.addAll(selectedDatanodes);
-            nodes.addAll(selectedDatanodes);
-            List<ReconstructECContainersCommand.DatanodeDetailsAndReplicaIndex>
-                    sourceDatanodesWithIndex = new ArrayList<>();
-            for (Pair<ContainerReplica, NodeStatus> src : sources.values()) {
-              sourceDatanodesWithIndex.add(
-                      new ReconstructECContainersCommand
-                              .DatanodeDetailsAndReplicaIndex(
-                              src.getLeft().getDatanodeDetails(),
-                              src.getLeft().getReplicaIndex()));
-            }
+      Map<Integer, Pair<ContainerReplica, NodeStatus>> sources =
+          filterSources(replicas, deletionInFlight);
+      List<DatanodeDetails> availableSourceNodes =
+          sources.values().stream().map(Pair::getLeft)
+              .map(ContainerReplica::getDatanodeDetails)
+              .filter(datanodeDetails ->
+                  datanodeDetails.getPersistedOpState() ==
+                      HddsProtos.NodeOperationalState.IN_SERVICE)
+              .collect(Collectors.toList());
 
-            final ReconstructECContainersCommand reconstructionCommand =
-                    new ReconstructECContainersCommand(id.getProtobuf().getId(),
-                            sourceDatanodesWithIndex, selectedDatanodes,
-                            int2byte(missingIndexes),
-                            repConfig);
-            // Keeping the first target node as coordinator.
-            commands.put(selectedDatanodes.get(0), reconstructionCommand);
-          }
-        } else {
-          LOG.warn("Cannot proceed for EC container reconstruction for {}, due"
-              + " to insufficient source replicas found. Number of source "
-              + "replicas needed: {}. Number of available source replicas are:"
-              + " {}. Available sources are: {}", container.containerID(),
-              repConfig.getData(), sources.size(), sources);
+      try {
+        processMissingIndexes(replicaCount, sources, availableSourceNodes,
+            excludedNodes, commands);
+        processDecommissioningIndexes(replicaCount, replicas,
+            availableSourceNodes, excludedNodes, commands);
+        processMaintenanceOnlyIndexes(replicaCount, replicas, excludedNodes,
+            commands);
+        // TODO - we should be able to catch SCMException here and check the
+        //        result code but the RackAware topology never sets the code.
+      } catch (CannotFindTargetsException e) {
+        // If we get here, we tried to find nodes to fix the under replication
```

Comment:
I am just thinking one corner case which could make both under and over replication processed at the same time?
Let's say, we have processed one underReplication already, but it;s not yet fully processed due to throttling. Now another replica lost and other replica with over replication. The second pass is still valid under replication and target choosing just failed for some reason? (What could be potential reasons? All nodes busy and if throttling logic respects that when implement later?). Then we hand over to over replication. This time over replication may be processed. SO, old under replication  task and new over replication task can be processed at same time possibly.

Since it is corner case, it is ok I think. Just to think about the case I am posting here.

(Are we stoping RM to process same container when one task is in progress for the same container?)

Reply:
We don't stop two operations on a container being in flight at once, but we do track the inflight add / delete ops on a replica. So if two over rep operations were scheduled on the same container, they will have taken account of the pending ops and should do no harm.

There is a change a delayed over rep command could impact an under-rep reconstruction by removing a replica it depends on, however the worst case is that the reconstruction fails and gets retried next time.

This whole Jira is solve an edge case of under and over rep on a small cluster. I feel we can live with these other cases too, as they should be rare in practice and I think will only delay containers reaching their ideal state - they will not actually do any harm.

Reply:
Got it. Anyway even if they both scheduled at same, they both mostly of different index, so that should not create any harm I beleive.